client/block.c

```diff
@@ -559,6 +640,50 @@ static int add_block_nolock(struct xdag_block *newBlock, xdag_time_t limit)
 		}
 	}
 
+#if CACHE == 1
+	if(g_xdag_extstats.cache_usage >= g_xdag_extstats.cache_size){
+		if (g_xdag_extstats.cache_hitrate<0.94 && g_xdag_extstats.cache_size*2 <= CACHE_MAX_SIZE){
+			if(!g_xdag_extstats.cache_size && CACHE_MAX_SIZE){
+				g_xdag_extstats.cache_size++;
+			}
+			else{
+				g_xdag_extstats.cache_size = g_xdag_extstats.cache_size*2;
+			}
+		}
+		else if(g_xdag_extstats.cache_hitrate>0.98 && !cache_miss && g_xdag_extstats.cache_size){
+			g_xdag_extstats.cache_size--;
+		}
+		for(int l=g_xdag_extstats.cache_usage;l>g_xdag_extstats.cache_size;l--){
+			if(cache_first != NULL){
+				struct cache_block* to_free = cache_first;
+				cache_first = cache_first->next;
+				if(cache_first == NULL){
+					cache_last = NULL;
+				}
+				ldus_rbtree_remove(&cache_root,&to_free->node);
+				free(to_free);
+				g_xdag_extstats.cache_usage--;
+				}
+				else{
+					break;
```

Comment:
I think it is wrong case is g_xdag_extstats.cache_usage > 0 but cache_first  is NULL. I would suggest to write warning to log here. So you can check if it really happens and why.